cpp/src/arrow/compute/exec/plan_test.cc

```diff
@@ -344,6 +344,47 @@ void TestSourceSink(
               Finishes(ResultWith(UnorderedElementsAreArray(exp_batches.batches))));
 }
 
+void TestRecordBatchReaderSourceSink(
+    std::function<Result<std::shared_ptr<RecordBatchReader>>(const BatchesWithSchema&)>
+        to_reader) {
+  ASSERT_OK_AND_ASSIGN(auto executor, arrow::internal::ThreadPool::Make(1));
+  ExecContext exec_context(default_memory_pool(), executor.get());
+  ASSERT_OK_AND_ASSIGN(auto plan, ExecPlan::Make(exec_context));
+  AsyncGenerator<std::optional<ExecBatch>> sink_gen;
+
+  auto exp_batches = MakeBasicBatches();
+  ASSERT_OK_AND_ASSIGN(std::shared_ptr<RecordBatchReader> reader, to_reader(exp_batches));
+
+  ASSERT_OK(Declaration::Sequence(
+                {
+                    {"record_batch_reader_source",
+                     RecordBatchReaderSourceNodeOptions{exp_batches.schema, reader}},
+                    {"sink", SinkNodeOptions{&sink_gen}},
+                })
+                .AddToPlan(plan.get()));
+
+  ASSERT_THAT(StartAndCollect(plan.get(), sink_gen),
+              Finishes(ResultWith(UnorderedElementsAreArray(exp_batches.batches))));
```

Comment:
Let's use `DeclarationToExecBatches` for new tests (feel free to fix existing tests too).

Reply:
Would you mind, if I do this in another PR for the rest of the items?

Reply:
I don't mind if you fix the other tests in a separate PR.  However, any newly added tests should follow the new format.  `TEST(ExecPlanExecution, SourceGroupedSum)` is a good example from this file that you can follow.  This test, for example, would become something like:

```
  auto exp_batches = MakeBasicBatches();
  ASSERT_OK_AND_ASSIGN(std::shared_ptr<RecordBatchReader> reader, to_reader(exp_batches));
  Declaration plan("record_batch_reader_source", RecordBatchReaderSourceNodeOptions{exp_batches.schema, reader});
  ASSERT_OK_AND_ASSIGN(auto out_batches, DeclarationToExecBatches(std::move(plan)));
  AssertExecBatchesEqualIgnoringOrder(exp_batches.schema, exp_batches.batches, out_batches);
```

Reply:
The follow up PR is WIP here: https://github.com/apache/arrow/pull/15288